src/connectors/sort-by/connectSortBy.ts

```diff
@@ -0,0 +1,189 @@
+});
+
+/**
+ * The **SortBy** connector provides the logic to build a custom widget that will display a
+ * list of indices. With Algolia, this is most commonly used for changing ranking strategy. This allows
+ * a user to change how the hits are being sorted.
+ */
+
+export type SortByItem = {
+  /**
+   * The name of the index to target.
+   */
+  value: string;
+  /**
+   * The label of the index to display.
+   */
+  label: string;
+};
+
+export type SortByConnectorParams = {
+  /**
+   * Array of objects defining the different indices to choose from.
+   */
+  items: SortByItem[];
+  /**
+   * Function to transform the items passed to the templates.
+   */
+  transformItems?: TransformItems<SortByItem>;
+};
+
+export type SortByRendererOptions = {
+  /**
+   * The initially selected index.
+   */
+  initialIndex?: string;
+  /**
+   * The currently selected index.
+   */
+  currentRefinement: string;
+  /**
+   * All the available indices
+   */
+  options: SortByItem[];
+  /**
+   * Switches indices and triggers a new search.
+   */
+  refine: (value: string) => void;
+  /**
+   * `true` if the last search contains no result.
+   */
+  hasNoResults: boolean;
+};
+
+export type SortByConnector = Connector<
+  SortByRendererOptions,
+  SortByConnectorParams
+>;
+
+const connectSortBy: SortByConnector = function connectSortBy(
+  renderFn,
+  unmountFn = noop
+) {
+  checkRendering(renderFn, withUsage());
+
+  const connectorState: ConnectorState = {};
+
+  type ConnectorState = {
+    setIndex?(indexName: string): void;
+    initialIndex?: string;
+  };
+
+  return widgetParams => {
+    const { items, transformItems = (x => x) as TransformItems<SortByItem> } =
+      widgetParams || {};
+
+    if (!Array.isArray(items)) {
+      throw new Error(
+        withUsage('The `items` option expects an array of objects.')
+      );
+    }
+
+    return {
+      $$type: 'ais.sortBy',
+
+      init(initOptions) {
+        const { instantSearchInstance } = initOptions;
+
+        const widgetRenderState = this.getWidgetRenderState(initOptions);
+        const currentIndex = widgetRenderState.currentRefinement;
+        const isCurrentIndexInItems = find(
+          items,
+          item => item.value === currentIndex
+        );
+
+        warning(
+          isCurrentIndexInItems !== undefined,
+          `The index named "${currentIndex}" is not listed in the \`items\` of \`sortBy\`.`
+        );
+
+        renderFn(
+          {
+            ...widgetRenderState,
+            instantSearchInstance,
+          },
+          true
+        );
+      },
+
+      render(renderOptions) {
+        const { instantSearchInstance } = renderOptions;
+        renderFn(
+          {
+            ...this.getWidgetRenderState(renderOptions),
+            instantSearchInstance,
+          },
+          false
+        );
+      },
+
+      dispose({ state }) {
+        unmountFn();
+
+        return state.setIndex(connectorState.initialIndex || '');
```

Comment:
we probably prefer throwing instead of setting to an empty string, or even `if initialIndex setIndex`

Reply:
Wouldn't it be a breaking change? For now, we accept it to be `undefined` if I understand correctly

Reply:
in the previous implementation setIndex was called with undefined, which I think would already not be the behaviour you want (it would make subsequent calls fail):

```

> const helper = require('algoliasearch-helper')
undefined
> const h = helper({search: console.log}, 'test')
undefined
> h.search()
[ { indexName: 'test', params: { facets: [], tagFilters: '' } } ]
> h.setIndex().search()
[ { indexName: undefined, params: { facets: [], tagFilters: '' } } ]
```

I'm in favour of either keeping it broken in the same was as before (`initialIndex!`), or only calling setIndex if we know the previous index, otherwise leaving it as-is

Reply:
Warns and only call `setIndex` when the previous index is defined: https://github.com/algolia/instantsearch.js/pull/4700/commits/f725991120710cd96605fc8a9c0db9fd1d04a3ea